lib/trino-filesystem-manager/src/main/java/io/trino/filesystem/manager/HdfsFileSystemLoader.java

```diff
@@ -0,0 +1,173 @@
+package io.trino.filesystem.manager;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.trino.filesystem.TrinoFileSystemFactory;
+import io.trino.spi.NodeManager;
+import io.trino.spi.Plugin;
+import jakarta.annotation.PreDestroy;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Streams.stream;
+import static java.nio.file.Files.newDirectoryStream;
+
+final class HdfsFileSystemLoader
+{
+    private final HdfsClassLoader classLoader;
+    private final Object manager;
+
+    public HdfsFileSystemLoader(
+            Map<String, String> config,
+            boolean azureEnabled,
+            boolean gcsEnabled,
+            boolean s3Enabled,
+            String catalogName,
+            NodeManager nodeManager,
+            OpenTelemetry openTelemetry)
+    {
+        Class<?> clazz = tryLoadExistingHdfsManager();
+
+        if (!getClass().getClassLoader().equals(Plugin.class.getClassLoader())) {
+            verify(clazz == null, "HDFS should not be on the classpath");
+            File sourceFile = getSourceFile();
+            File directory;
+            if (sourceFile.isDirectory()) {
+                verify(sourceFile.getPath().endsWith("/target/classes"), "Source file not in 'target' directory: %s", sourceFile);
+                directory = new File(sourceFile.getParentFile().getParentFile().getParentFile(), "trino-hdfs/target/hdfs");
+            }
+            else {
+                directory = new File(sourceFile.getParentFile(), "hdfs");
+            }
+            verify(directory.isDirectory(), "HDFS directory is missing: %s", directory);
+            classLoader = createClassLoader(directory);
+            clazz = loadHdfsManager(classLoader);
+        }
+        else {
+            verify(clazz != null, "HDFS should be on the classpath for tests");
+            classLoader = null;
+        }
+
+        try {
+            manager = clazz.getConstructor(Map.class, boolean.class, boolean.class, boolean.class, String.class, NodeManager.class, OpenTelemetry.class)
+                    .newInstance(config, azureEnabled, gcsEnabled, s3Enabled, catalogName, nodeManager, openTelemetry);
+        }
+        catch (ReflectiveOperationException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public Set<String> configure()
+    {
+        try {
+            return (Set<String>) manager.getClass().getMethod("configure").invoke(manager);
+        }
+        catch (ReflectiveOperationException e) {
+            throw new RuntimeException("Failed to configure HDFS:\n<<<<<\n%s\n>>>>>".formatted(e.getCause()), e);
```

Comment:
The reason to do this is to that the error from the HDFS Bootstrap is at the top of the error message, rather than hidden at the bottom. If you don't do that, all you get is `Message: InvocationTargetException` and some misleading errors about unused configuration properties (which were actually used by the HDFS code).

Example error with plain rethrowing using `throw new RuntimeException(e)`:

```
io.airlift.bootstrap.ApplicationConfigurationException: Configuration errors:

1) Error: An exception was caught and reported. Message: InvocationTargetException
  at ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)

2) Error: Configuration property 'hive.cache.enabled' was not used

3) Error: Configuration property 'hive.cache.location' was not used

4) Error: Configuration property 'hive.s3.security-mapping.config-file' was not used

4 errors

======================
Full classname legend:
======================
ConfigurationFactory:      "io.airlift.configuration.ConfigurationFactory"
InvocationTargetException: "java.lang.reflect.InvocationTargetException"
========================
End of classname legend:
========================

	at io.airlift.bootstrap.Bootstrap.configure(Bootstrap.java:217)
	at io.airlift.bootstrap.Bootstrap.initialize(Bootstrap.java:246)
	at io.trino.plugin.hive.InternalHiveConnectorFactory.createConnector(InternalHiveConnectorFactory.java:134)
	at io.trino.plugin.hive.InternalHiveConnectorFactory.createConnector(InternalHiveConnectorFactory.java:82)
	at io.trino.plugin.hive.HiveConnectorFactory.create(HiveConnectorFactory.java:53)
	at io.trino.plugin.hive.TestHivePlugin.lambda$testS3SecurityMappingAndHiveCachingMutuallyExclusive$1(TestHivePlugin.java:163)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:63)
	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:892)
	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1366)
	at org.assertj.core.api.Assertions.assertThatThrownBy(Assertions.java:1210)
	at io.trino.plugin.hive.TestHivePlugin.testS3SecurityMappingAndHiveCachingMutuallyExclusive(TestHivePlugin.java:161)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:105)
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at io.trino.filesystem.manager.HdfsFileSystemLoader.configure(HdfsFileSystemLoader.java:93)
	at io.trino.filesystem.manager.FileSystemModule.setup(FileSystemModule.java:71)
	at io.airlift.configuration.AbstractConfigurationAwareModule.configure(AbstractConfigurationAwareModule.java:48)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
	at com.google.inject.spi.Elements.getElements(Elements.java:113)
	at com.google.inject.spi.Elements.getElements(Elements.java:106)
	at io.airlift.configuration.ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)
	at io.airlift.bootstrap.Bootstrap.configure(Bootstrap.java:202)
	... 32 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.trino.filesystem.manager.HdfsFileSystemLoader.configure(HdfsFileSystemLoader.java:90)
	... 39 more
Caused by: io.airlift.bootstrap.ApplicationConfigurationException: Configuration errors:

1) Error: An exception was caught and reported. Message: S3 security mapping is not compatible with Hive caching
  at ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)

1 error

======================
Full classname legend:
======================
ConfigurationFactory: "io.airlift.configuration.ConfigurationFactory"
========================
End of classname legend:
========================

	at io.airlift.bootstrap.Bootstrap.configure(Bootstrap.java:217)
	at io.trino.filesystem.hdfs.HdfsFileSystemManager.configure(HdfsFileSystemManager.java:87)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 41 more
Caused by: java.lang.IllegalArgumentException: S3 security mapping is not compatible with Hive caching
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:143)
	at io.trino.hdfs.s3.HiveS3Module.bindSecurityMapping(HiveS3Module.java:86)
	at io.trino.hdfs.s3.HiveS3Module.setup(HiveS3Module.java:44)
	at io.airlift.configuration.AbstractConfigurationAwareModule.configure(AbstractConfigurationAwareModule.java:48)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
	at com.google.inject.spi.Elements.getElements(Elements.java:113)
	at com.google.inject.spi.Elements.getElements(Elements.java:106)
	at io.airlift.configuration.ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)
	at io.airlift.bootstrap.Bootstrap.configure(Bootstrap.java:202)
	... 43 more
```

Example with the cause (with increased marker width):

```
io.airlift.bootstrap.ApplicationConfigurationException: Configuration errors:

1) Error: An exception was caught and reported. Message: Failed to configure HDFS:
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
ApplicationConfigurationException: Configuration errors:

1) Error: An exception was caught and reported. Message: S3 security mapping is not compatible with Hive caching
  at ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)

1 error

======================
Full classname legend:
======================
ConfigurationFactory: "io.airlift.configuration.ConfigurationFactory"
========================
End of classname legend:
========================

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
  at ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)

2) Error: Configuration property 'hive.cache.enabled' was not used

3) Error: Configuration property 'hive.cache.location' was not used

4) Error: Configuration property 'hive.s3.security-mapping.config-file' was not used

4 errors

======================
Full classname legend:
======================
ApplicationConfigurationException: "io.airlift.bootstrap.ApplicationConfigurationException"
ConfigurationFactory:              "io.airlift.configuration.ConfigurationFactory"
========================
End of classname legend:
========================

	at io.airlift.bootstrap.Bootstrap.configure(Bootstrap.java:217)
	at io.airlift.bootstrap.Bootstrap.initialize(Bootstrap.java:246)
	at io.trino.plugin.hive.InternalHiveConnectorFactory.createConnector(InternalHiveConnectorFactory.java:134)
	at io.trino.plugin.hive.InternalHiveConnectorFactory.createConnector(InternalHiveConnectorFactory.java:82)
	at io.trino.plugin.hive.HiveConnectorFactory.create(HiveConnectorFactory.java:53)
	at io.trino.plugin.hive.TestHivePlugin.lambda$testS3SecurityMappingAndHiveCachingMutuallyExclusive$1(TestHivePlugin.java:163)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:63)
	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:892)
	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1366)
	at org.assertj.core.api.Assertions.assertThatThrownBy(Assertions.java:1210)
	at io.trino.plugin.hive.TestHivePlugin.testS3SecurityMappingAndHiveCachingMutuallyExclusive(TestHivePlugin.java:161)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:105)
Caused by: java.lang.RuntimeException: Failed to configure HDFS:
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
io.airlift.bootstrap.ApplicationConfigurationException: Configuration errors:

1) Error: An exception was caught and reported. Message: S3 security mapping is not compatible with Hive caching
  at ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)

1 error

======================
Full classname legend:
======================
ConfigurationFactory: "io.airlift.configuration.ConfigurationFactory"
========================
End of classname legend:
========================

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
	at io.trino.filesystem.manager.HdfsFileSystemLoader.configure(HdfsFileSystemLoader.java:93)
	at io.trino.filesystem.manager.FileSystemModule.setup(FileSystemModule.java:71)
	at io.airlift.configuration.AbstractConfigurationAwareModule.configure(AbstractConfigurationAwareModule.java:48)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
	at com.google.inject.spi.Elements.getElements(Elements.java:113)
	at com.google.inject.spi.Elements.getElements(Elements.java:106)
	at io.airlift.configuration.ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)
	at io.airlift.bootstrap.Bootstrap.configure(Bootstrap.java:202)
	... 32 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.trino.filesystem.manager.HdfsFileSystemLoader.configure(HdfsFileSystemLoader.java:90)
	... 39 more
Caused by: io.airlift.bootstrap.ApplicationConfigurationException: Configuration errors:

1) Error: An exception was caught and reported. Message: S3 security mapping is not compatible with Hive caching
  at ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)

1 error

======================
Full classname legend:
======================
ConfigurationFactory: "io.airlift.configuration.ConfigurationFactory"
========================
End of classname legend:
========================

	at io.airlift.bootstrap.Bootstrap.configure(Bootstrap.java:217)
	at io.trino.filesystem.hdfs.HdfsFileSystemManager.configure(HdfsFileSystemManager.java:87)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 41 more
Caused by: java.lang.IllegalArgumentException: S3 security mapping is not compatible with Hive caching
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:143)
	at io.trino.hdfs.s3.HiveS3Module.bindSecurityMapping(HiveS3Module.java:86)
	at io.trino.hdfs.s3.HiveS3Module.setup(HiveS3Module.java:44)
	at io.airlift.configuration.AbstractConfigurationAwareModule.configure(AbstractConfigurationAwareModule.java:48)
	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
	at com.google.inject.spi.Elements.getElements(Elements.java:113)
	at com.google.inject.spi.Elements.getElements(Elements.java:106)
	at io.airlift.configuration.ConfigurationFactory.registerConfigurationClasses(ConfigurationFactory.java:186)
	at io.airlift.bootstrap.Bootstrap.configure(Bootstrap.java:202)
	... 43 more
```

Reply:
Let me see if I can clean this up by catching `ApplicationConfigurationException` and formatting it differently, like we do for the main server.